usage/src/org/apache/cloudstack/quota/QuotaManagerImpl.java

```diff
@@ -0,0 +1,397 @@
+                    usageRecord.getStartDate(), usageRecord.getEndDate());
+            _quotaUsageDao.persist(quota_usage);
+            quotalist.add(quota_usage);
+        }
+        tariff = _quotaTariffDao.findTariffPlanByUsageType(QuotaTypes.CPU_CLOCK_RATE, usageRecord.getEndDate());
+        if (tariff != null && !tariff.getCurrencyValue().equals(0)) {
+            BigDecimal speed = new BigDecimal(serviceoffering.getSpeed() / 100.00);
+            onehourcostper100mhz = tariff.getCurrencyValue().multiply(aggregationRatio);
+            speedquotausage = rawusage.multiply(onehourcostper100mhz).multiply(speed);
+            quota_usage = new QuotaUsageVO(usageRecord.getId(), usageRecord.getZoneId(), usageRecord.getAccountId(), usageRecord.getDomainId(), QuotaTypes.CPU_CLOCK_RATE, speedquotausage,
+                    usageRecord.getStartDate(), usageRecord.getEndDate());
+            _quotaUsageDao.persist(quota_usage);
+            quotalist.add(quota_usage);
+        }
+        tariff = _quotaTariffDao.findTariffPlanByUsageType(QuotaTypes.MEMORY, usageRecord.getEndDate());
+        if (tariff != null && !tariff.getCurrencyValue().equals(0)) {
+            BigDecimal memory = new BigDecimal(serviceoffering.getRamSize());
+            onehourcostper1mb = tariff.getCurrencyValue().multiply(aggregationRatio);
+            memoryquotausage = rawusage.multiply(onehourcostper1mb).multiply(memory);
+            quota_usage = new QuotaUsageVO(usageRecord.getId(), usageRecord.getZoneId(), usageRecord.getAccountId(), usageRecord.getDomainId(), QuotaTypes.MEMORY, memoryquotausage,
+                    usageRecord.getStartDate(), usageRecord.getEndDate());
+            _quotaUsageDao.persist(quota_usage);
+            quotalist.add(quota_usage);
+        }
+        tariff = _quotaTariffDao.findTariffPlanByUsageType(QuotaTypes.RUNNING_VM, usageRecord.getEndDate());
+        if (tariff != null && !tariff.getCurrencyValue().equals(0)) {
+            onehourcostforvmusage = tariff.getCurrencyValue().multiply(aggregationRatio);
+            vmusage = rawusage.multiply(onehourcostforvmusage);
+            quota_usage = new QuotaUsageVO(usageRecord.getId(), usageRecord.getZoneId(), usageRecord.getAccountId(), usageRecord.getDomainId(), QuotaTypes.RUNNING_VM, vmusage,
+                    usageRecord.getStartDate(), usageRecord.getEndDate());
+            _quotaUsageDao.persist(quota_usage);
+            quotalist.add(quota_usage);
+        }
+
+        usageRecord.setQuotaCalculated(1);
+        _usageDao.persist(usageRecord);
+        return quotalist;
+    }
+
+    @DB
+    public QuotaUsageVO updateQuotaAllocatedVMUsage(UsageVO usageRecord, final BigDecimal aggregationRatio) {
+        QuotaUsageVO quota_usage = null;
+        QuotaTariffVO tariff = _quotaTariffDao.findTariffPlanByUsageType(QuotaTypes.ALLOCATED_VM, usageRecord.getEndDate());
+        if (tariff != null && !tariff.getCurrencyValue().equals(0)) {
```

Comment:
The code only checks if tariff is not set or is 0. In which case it does not do the calculation as there will be no usage. There is no impact on + or - tariff.